src/miner.erl

```diff
@@ -750,75 +801,166 @@ common_enough_or_default(Threshold, Xs, Default) ->
-                           BlockTime;
-                       _ ->
-                           (LastBlockTimestamp - StartBlockTime) / max(Height - ActualStartHeight, 1)
-                   end,
-    BlockTimeDeviation0 = BlockTime - AvgBlockTime,
-    lager:info("average ~p block times ~p difference ~p", [Height, AvgBlockTime, BlockTime - AvgBlockTime]),
+        undefined ->
+            BlockTime;
+        _ ->
+            (LastBlockTimestamp - StartBlockTime) / BlockRange
+    end,
+    {AvgBlockTime, BlockRange}.
+
+get_next_block_time({AvgBlockTime, BlockRange}, BlockTime, LastBlockTimestamp) ->
+    Now = erlang:system_time(seconds),
+    DifferenceInTime = (BlockTime - AvgBlockTime) * BlockRange,
     BlockTimeDeviation =
-        case BlockTimeDeviation0 of
+        case DifferenceInTime of
             N when N > 0 ->
                 min(1, catchup_time(abs(N)));
             N ->
                 -1 * catchup_time(abs(N))
         end,
-    NextBlockTime = max(0, (LastBlockTimestamp + BlockTime + BlockTimeDeviation) - Now),
-    lager:info("Next block after ~p is in ~p seconds", [LastBlockTimestamp, NextBlockTime]),
-    Timer = erlang:send_after(NextBlockTime * 1000, self(), block_timeout),
-
-    %% now figure out the late block timer
-    erlang:cancel_timer(State#state.late_block_timer),
-    LateBlockTimeout = application:get_env(miner, late_block_timeout_seconds, 120),
-    LateTimer = erlang:send_after((LateBlockTimeout + NextBlockTime) * 1000, self(), late_block_timeout),
-
-    State#state{block_timer=Timer, late_block_timer = LateTimer}.
+    %% if chain has been stopped longer then LastBlockTimeout prevent 0 NextBlockTime
+    max(Now, LastBlockTimestamp) + BlockTime + BlockTimeDeviation.
 
 %% input in fractional seconds, the number of seconds between the
 %% target block time and the average total time over the target period
 %% output in seconds of adjustment to apply to the block time target
 
-%% the constants here are by feel: currently at 100k target and 5sec
-%% adjustment it takes roughly a day of blocks to make up a tenth of a
-%% second. with the new 50k target we can expect double the adjustment
-%% leverage, so 1s adjustments will take roughly a day to make up the
-%% final 0.04 (twice as much leverage, but 20% of the rate).
-
 %% when drift is small or 0, let it accumulate for a bit
 catchup_time(N) when N < 0.001 ->
     0;
-%% when it's still relatively small, apply gentle adjustments
-catchup_time(N) when N < 0.01 ->
-    1;
-%% if it's large, jam on the gas
-catchup_time(_) ->
-    10.
+%% try and catch up within 10 blocks, max 10 seconds
+catchup_time(N) ->
+    min(10, ceil(N / 10)).
+
+
+%% ------------------------------------------------------------------
+%% EUNIT Tests
+%% ------------------------------------------------------------------
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+
+%% Confirm changes to make catchup_time proportional
+catchup_time_test() ->
```

Comment:
these unit test where specific to my changes in #1557 . Remove if not using that logic or they will not work.

Reply:
Oh yea, haven't gotten to "tests" yet. But figured I'd post this to get some more eyes on it.